feat(security): Add freeze guard protection against pre-freeze timelocked transactions

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/419

Fixes [ENG-1274](https://linear.app/decent-labs/issue/ENG-1274/implement-security-protection-against-pre-freeze-timelocked)

Implements critical security features to prevent circumvention of freeze mechanism:

Security Features:
- Prevent new timelocks when DAO is frozen
- Block execution of transactions timelocked before the most recent freeze
- Track permanent freeze timestamp that is never cleared

Changes:
- Add lastKnownFreezeTime() to IFreezable interface
- Implement lastFreezeTimestamp tracking in FreezeVotingBase
- Add security checks in FreezeGuardMultisigV1:
  - timelockTransaction() reverts if DAO is frozen
  - checkTransaction() reverts if transaction was timelocked before freeze
- Add new TimelockedBeforeFreeze error for clarity
- Update MockFreezable to support lastKnownFreezeTime()
- Add comprehensive tests for security invariants

SECURITY INVARIANT: Any transaction timelocked before the most recent freeze is permanently invalidated and can never be executed, even after the DAO is unfrozen. This prevents malicious actors from queuing harmful transactions and waiting for an unfreeze to execute them.

The lastFreezeTimestamp is set when a DAO freezes and is NEVER cleared, ensuring the security model remains intact across freeze/unfreeze cycles.

tbwebb22 commented on contracts/deployables/freeze-voting/FreezeVotingBase.sol:
```diff
@@ -135,6 +137,23 @@ abstract contract FreezeVotingBase is
+     * This timestamp is NEVER cleared, even after unfreeze or auto-expiry.
+     * Used by guards to permanently invalidate all transactions timelocked before this time.
+     */
+    function lastKnownFreezeTime()
```
Can we rename this to `lastFreezeTime()` ?

adamgall replied:
good call, updated

adamgall replied:
@tbwebb22 can you provide another review please?

adamgall committed:
feat(security): Add freeze guard protection against pre-freeze timelocked transactions

Implements critical security features to prevent circumvention of freeze mechanism:

Security Features:
- Prevent new timelocks when DAO is frozen
- Block execution of transactions timelocked before the most recent freeze
- Track permanent freeze timestamp that is never cleared

Changes:
- Add lastFreezeTime() to IFreezable interface
- Implement lastFreezeTimestamp tracking in FreezeVotingBase
- Add security checks in FreezeGuardMultisigV1:
  - timelockTransaction() reverts if DAO is frozen
  - checkTransaction() reverts if transaction was timelocked before freeze
- Add new TimelockedBeforeFreeze error for clarity
- Update MockFreezable to support lastKnownFreezeTime()
- Add comprehensive tests for security invariants

SECURITY INVARIANT: Any transaction timelocked before the most recent
freeze is permanently invalidated and can never be executed, even after
the DAO is unfrozen. This prevents malicious actors from queuing harmful
transactions and waiting for an unfreeze to execute them.

The lastFreezeTimestamp is set when a DAO freezes and is NEVER cleared,
ensuring the security model remains intact across freeze/unfreeze cycles.
